### PR TITLE
Interpolation for stationary obstacles.

### PR DESCRIPTION
**Pull request by yogishadixit**
For #260 

Got rid of extrapolation as discussed the telemetry rate requirement will take care of it.

### CODE REVIEW

File: server/auvsi_suas/models/stationary_obstacle.py [hunk 1]
```diff
@@ -23,6 +28,87 @@ def __unicode__(self):
...
+            return False
+
+        rm = units.feet_to_meters(self.cylinder_radius)
+        hm = units.feet_to_meters(self.cylinder_height)
+
+        # Calculate equation of line and substitute into circle equation.
+        # Equation of obstacle circle:
+        # (x - latc)^2 + (y - laty)^2 = self.cylinder_radius^2
+        delta_x = x2 - x1
+        if delta_x == 0:
```
Comment (pmtischler) [inline server/auvsi_suas/models/stationary_obstacle.py]: Inline delta_x, it's only used once

File: server/auvsi_suas/models/stationary_obstacle.py [hunk 2]
```diff
@@ -52,9 +138,20 @@ def evaluate_collision_with_uas(self, uas_telemetry_logs):
...
+        zone, north = distance.utm_zone(self.gps_position.latitude,
+                                        self.gps_position.longitude)
+        utm = distance.proj_utm(zone, north)
+        for i in range(0, len(uas_telemetry_logs)):
+            cur_log = uas_telemetry_logs[i]
+            if i > 0:
+                cur_log = uas_telemetry_logs[i]
+                prev_log = uas_telemetry_logs[i - 1]
+                if self.determine_interpolated_collision(
+                        uas_telemetry_logs[i - 1], uas_telemetry_logs[i], utm):
```
Comment (pmtischler) [inline server/auvsi_suas/models/stationary_obstacle.py]: Doesn't use cur_log or prev_log

File: server/auvsi_suas/models/stationary_obstacle.py [hunk 3]
```diff
@@ -52,9 +138,20 @@ def evaluate_collision_with_uas(self, uas_telemetry_logs):
             Whether a UAS telemetry log reported indicates a collision with the
             obstacle.
         """
-        for cur_log in uas_telemetry_logs:
-            if self.contains_pos(cur_log.uas_position):
-                return True
+        zone, north = distance.utm_zone(self.gps_position.latitude,
+                                        self.gps_position.longitude)
+        utm = distance.proj_utm(zone, north)
+        for i in range(0, len(uas_telemetry_logs)):
```
Comment (pmtischler) [inline server/auvsi_suas/models/stationary_obstacle.py]: Use enumerate

File: server/auvsi_suas/models/stationary_obstacle.py [hunk 4]
```diff
@@ -23,6 +28,87 @@ def __unicode__(self):
...
+            m = (y2 - y1) / (x2 - x1)
+            b = y1 - m * x1
+            # Substitute in line equation and solve for x.
+            p = [m**2 + 1, (2 * m * (b - cy)) - (2 * cx),
+                 cx**2 + (b - cy)**2 - rm**2]
+            roots = np.roots(p)
+            for root in roots:
+                # Solve for altitude and check if within bounds.
+                zcalc = ((root - x1) * (z2 - z1) / (x2 - x1)) + z1
+                if np.isreal(root) and zcalc <= hm:
```
Comment (pmtischler) [inline server/auvsi_suas/models/stationary_obstacle.py]: While you're here, please fix my bug in contain_pos to not check if alt < 0.

File: server/auvsi_suas/models/stationary_obstacle_test.py [hunk 5]
```diff
@@ -33,16 +35,28 @@
      (-76.0001, 38.0001, 0),
      (-76.0001, 38.0001, 100)],
     # Outside positions
-    [(-76.001, 38, 50),
```
Comment (pmtischler) [inline server/auvsi_suas/models/stationary_obstacle_test.py]: Why did this change?

File: server/auvsi_suas/models/stationary_obstacle_test.py [hunk 6]
```diff
@@ -53,6 +67,39 @@ def test_unicode(self):
         obst.save()
         self.assertTrue(obst.__unicode__())
 
+    def create_uas_logs(self, user, entries):
```
Comment (pmtischler) [inline server/auvsi_suas/models/stationary_obstacle_test.py]: Do we have a test utility for this in another file?

File: server/auvsi_suas/models/stationary_obstacle_test.py [hunk 7]
```diff
@@ -33,16 +35,28 @@
...
+    [(-76.001, 38, 150),
+     (-76, 38.002, 150),
+     (-76, 38, 150),
      (-76, 38, 101)]
 )  # yapf: disable
 
+# (lat, lon, rad, height)
+TESTDATA_STATOBST_INTERP_OBS = (38.145146, -76.427522, 30, 100)
+# (lat, lon, alt)
+TESTDATA_STATOBST_INTERP_TELEM = [
```
Comment (pmtischler) [inline server/auvsi_suas/models/stationary_obstacle_test.py]: Some constants are here, others are in the function. Please make it one or the other.

Comment (pmtischler) [inline server/auvsi_suas/models/stationary_obstacle_test.py]: Can you add a few more test cases? E.g. altitude.

File: server/auvsi_suas/models/stationary_obstacle.py [hunk 8]
```diff
@@ -52,9 +138,20 @@ def evaluate_collision_with_uas(self, uas_telemetry_logs):
...
+        for i in range(0, len(uas_telemetry_logs)):
+            cur_log = uas_telemetry_logs[i]
+            if i > 0:
+                cur_log = uas_telemetry_logs[i]
+                prev_log = uas_telemetry_logs[i - 1]
+                if self.determine_interpolated_collision(
+                        uas_telemetry_logs[i - 1], uas_telemetry_logs[i], utm):
+                    return True
+            else:
+                if self.contains_pos(cur_log.uas_position):
```
Comment (pmtischler) [inline server/auvsi_suas/models/stationary_obstacle.py]: Can we always run this simple check in addition to interpolated?

File: server/auvsi_suas/models/stationary_obstacle.py (hunk 4 above)
Comment (yogishadixit) [inline server/auvsi_suas/models/stationary_obstacle.py]: Wait, so change
```
if (aerial_alt < 0 or aerial_alt > self.cylinder_height):
```
to 
```
if aerial_alt > self.cylinder_height:
```
Is the negative altitude handled somewhere else?

**Comment by pmtischler** [comment]
Stationary obstacles are (-inf, height]. The height is defined in feet ASL,
not AGL.

File: server/auvsi_suas/models/stationary_obstacle_test.py (hunk 5 above)
Comment (yogishadixit) [inline server/auvsi_suas/models/stationary_obstacle_test.py]: Oops, didn't mean to change this line as well, reverted.

File: server/auvsi_suas/models/stationary_obstacle_test.py (hunk 6 above)
Comment (yogishadixit) [inline server/auvsi_suas/models/stationary_obstacle_test.py]: We have it defined in other tests, but not in a test utility file. Maybe that can be addressed in a follow up PR.

**Pull request merged**